sw/device/silicon_creator/lib/irq_asm_functest.c

```diff
@@ -0,0 +1,80 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
+#include "sw/device/silicon_creator/lib/drivers/rstmgr.h"
+#include "sw/device/silicon_creator/lib/drivers/watchdog.h"
+#include "sw/device/silicon_creator/lib/error.h"
+#include "sw/device/silicon_creator/lib/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "rstmgr_regs.h"
+
+// The reset reason value is really a bitfield. The power-on-reset indicator
+// is defined by rstmgr_regs.h.
+#define RESET_REASON_POR (1 << RSTMGR_RESET_INFO_POR_BIT)
+// FIXME: I don't know where the HW_REQ field of the reset reason register
+// is defined.  I observe a value of 2 for a watchdog reset.
+#define RESET_REASON_WATCHDOG \
+  ((2 << RSTMGR_RESET_INFO_HW_REQ_OFFSET) | RESET_REASON_POR)
+
+void *const kRetentionRamBase = (void *)TOP_EARLGREY_RAM_RET_AON_BASE_ADDR;
+const size_t kRetentionRamSize = TOP_EARLGREY_RAM_RET_AON_SIZE_BYTES;
+
+extern void _asm_exception_handler(void);
+
+const test_config_t kTestConfig;
+
+// The test phases are tracked in retention RAM so that we ensure the reset
+// happened in the correct phase of the test.
+typedef enum TestPhase {
+  kTestPhaseInit = 0,
+  kTestPhaseReset = 1,
+  kTestPhaseDone = 2,
+} test_phase_t;
+
+bool test_main(void) {
+  rom_error_t result = kErrorOk;
+  uint32_t reason = rstmgr_reason_get();
+  rstmgr_alert_info_enable();
+  LOG_INFO("reset_info = %08x", reason);
+
+  volatile test_phase_t *phase = (volatile test_phase_t *)kRetentionRamBase;
+  if (reason == RESET_REASON_POR) {
+    // Power-on: zero out the retention RAM.
+    memset(kRetentionRamBase, 0, kRetentionRamSize);
+
+    LOG_INFO("Calling exception handler to reset device.");
+    *phase = kTestPhaseReset;
+    _asm_exception_handler();
+
+    // We should never get here - the exception handler should cause a reset
+    // and we should see a reset reason of 0x11.
+    *phase = kTestPhaseDone;
+    result = kErrorUnknown;
+    LOG_ERROR("Test failure: should have reset before this line.");
+  } else if (reason == RESET_REASON_WATCHDOG) {
+    LOG_INFO("Detected reset after exception test");
+    if (*phase != kTestPhaseReset) {
+      LOG_ERROR("Test failure: expected phase %d but got phase %d",
+                kTestPhaseReset, *phase);
+      result = kErrorUnknown;
+    } else {
+      result = kErrorOk;
+    }
```

Comment:
This line is statically unneeded, right?

Reply:
I agree. IMHO initializing `result` to `kErrorUnknown` and setting it to `kErrorOk` only here would be better.

Reply:
I've changed it to return true and false in the right places.